src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php

```diff
@@ -160,7 +160,9 @@ protected static function ensureKernelShutdown()
             static::$kernel->shutdown();
             static::$booted = false;
 
-            if ($container instanceof ResetInterface) {
+            if ($container->has('services_resetter')) {
+                $container->get('services_resetter')->reset();
+            } elseif ($container instanceof ResetInterface) {
                 $container->reset();
```

Comment:
`$container->reset()` must still be called, to reset the instances hold by the container itself (see the code of line 302).

Btw, if you ensure that the `services_resetter` service is instantiated, the container reset call will then call it (as ServiceResetter implements the ResetInterface).

The reason it was not called is because `Container::reset` only resets *instantiated* resettable services

Reply:
@stof Fixed, see https://github.com/symfony/symfony/compare/549ada5257c771da92d51aff76e224e1ac5d31f4..5491aa54a3205acfaf5ba62c134bddc26d794a36

Initially, I wanted to just instantiate `services_resetter` and let `Container::reset()` call reset on it, but that didn’t work because the `services_resetter` only resets instantiated services. Because of line 302, by the time it’s called, there aren’t any instantiated services left.

The downside of the current approach is that some services will get reset twice, but I don’t see a better solution at the moment.

Reply:
ah indeed. The container resets itself before resetting the services.

I think this might actually be an issue if the `reset` method of a service triggers the instantiation of some other service. I think the resetting of the container state should probably be moved after the loop.

Reply:
@stof Done https://github.com/symfony/symfony/compare/5491aa54a3205acfaf5ba62c134bddc26d794a36..53d0f56be6a315055cb3ceef09a7f6bcb233a120